keps/sig-apps/4040-controller-name/README.md

```diff
@@ -0,0 +1,371 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This feature unblocks custom handling for selected standard
+Kubernetes API objects, while still allowing the default
+behavior for other API objects. For now it focuses only on Job and HPA, but other controllers can reuse this pattern.
+
+## Motivation
+
+Currently Kubernetes users, who need custom behavior for
+some API objects, can either:
+
+* disable the corresponding controller in controller-manager
+and implement a new one that provides both modified(for marked
+objects) and original(for all other) functionality.
+
+* duplicate/fork the API as CRD.
+
+However, on managed environments (like, for example, GKE),
+only the second option is available as cloud providers often
+don't allow to modify `controller-manager` configuration.
+
+### Goals
+
+* Establish a simple and common method to mark an API object as
+to be handled by a non default controller, in a similar fashion
+as pods can specify their custom scheduler.
+
+* Apply it and provide implementation for Job and HPA.
+
+### Non-Goals
+
+* Add controller name to all possible K8S objects at once.
+
+
+* While the KEP enables providing custom handling for standard
+K8S APIs,
+it doesn't focus on any customization or additional
+configuration that might be possible using it. Additional
+configuration options may be subject for the following
+enhancement proposals.
+
+
+## Proposal
+
+Add `controllerName` at the top level of Spec of API objects to
+specify which controller should handle them. If the field
+value is `”default”` or the field is not specified at all,
+then the default controller from controller manager handles the
+object in the usual way. If the specified  value is different,
+then the controller, which identifies itself with the given name, should handle the object and all others ignore it.
+
+Initially add the field only to Job and HPA. If the feedback is good, consider expansion to other objects.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to run a modified, custom version of  ControllerManager in
+cluster and handle some (not all) of the Jobs with it.
+
+#### Story 2
+
+I want to provide a significantly different implementation of
```

Comment:
I buy this story.  You could use annotations for this, which we did in Service for a while and ultimately added a field for it because annotations suck.  That said, we half-assed it and DIDN'T add a class resource, because we know that Gateway is the real path forward (obviates much of Service API).  

So I actually buy `HPAClass` as an idea, though I wonder about how much the API will need to evolve to provide a "significantly different implementation".  Job is less obvious to me.